src/util.rs

```diff
@@ -472,7 +492,58 @@ pub(crate) mod ptr {
         }
     }
 
-    impl<'a, T: 'a + ?Sized> Debug for Ptr<'a, T> {
+    impl<'a, T: 'a + ?Sized, const ASSUME_ALIGNED: bool> Debug for Ptr<'a, T, ASSUME_ALIGNED> {
+        #[inline]
+        fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+            self.ptr.fmt(f)
+        }
+    }
+
+    /// A reference to a validly-initialized `T` of lifetime `'a` that might
+    /// *not* be validly aligned for `T`.
+    #[doc(hidden)]
+    #[allow(unreachable_pub)] // false positive?
+    pub struct MaybeAligned<'a, T: ?Sized> {
+        /// INVARIANTS:
+        /// 1. `ptr` is validly-initialized for `T`
+        ptr: Ptr<'a, T, false>,
+    }
+
+    impl<'a, T> MaybeAligned<'a, T>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Read the value from `MaybeAligned`.
+        ///
+        /// This is only available if `T` is [`Copy`].
+        #[inline]
+        pub fn read_unaligned(self) -> T
+        where
+            T: Copy,
+        {
+            let raw = self.ptr.ptr.as_ptr();
+            // SAFETY: By invariant on `MaybeAligned`, `raw` contains
+            // validly-initialized data for `T`. The value is safe to read and
+            // return, because `T` is copy.
+            unsafe { core::ptr::read_unaligned(raw) }
```

Comment:
It might be nice to add a `read` method to `Ptr`. That'd give us a more local spot to write down the safety preconditions on that method and then to justify the implementation in terms of `Ptr`'s internal invariants. It'd also presumably mean that the method's safety preconditions were pretty minimal.

Reply:
Now that `MaybeAligned` is just an alias on `Ptr`, this is technically now the case! However, the invariant bounds are overly restrictive.  We currently define `read_unaligned` on `invariant::Shared`, and we could also define it on `invariant::Unique` — but what about `invariant::AnyAliasing`?

Reply:
Can you leave a TODO comment on `read_unaligned` to consider loosening the requirements at some point in the future? In particular, I could see us introducing a trait which represents "at least shared" aliasing, implementing it for both `invariant::Shared` and `invariant::Exclusive`, and using that trait (rather than a particular concrete type) as the aliasing bound.